File: GRACE/aux_grace/getAUC.R
getAUC <- function(dps_test, densities, labels, classes){
  # ============================================================
  # Project:    Disease progression modeling from early AD stage
  # Repository: https://github.com/cplatero/preAD_DPM
  # Author:     Jorge Bengoa
  # Email:      j.bpinedo@alumnos.upm.es
  # Institution:Universidad Politécnica de Madrid
  # ------------------------------------------------------------
  # Filename:    getAUC.R
  # Description: Script for evaluating the GRACE model (AUC)
  #              on longitudinal neuropsychological data.
  #
  # Version:    1.0
  # Date:       2025-05-09
  # Requires:   grace
  # ============================================================
  require(pROC)
  #dps_test <- unlist(dps_test)
  dps_test <- as.vector(dps_test)
  #labels <- unlist(labels)
  labels <- as.vector(labels)
  mask <- is.na(labels)
  #dps_test <- dps_test[!mask]
  labels <- labels[!mask]
  dps_test <- dps_test[!is.na(dps_test)]
  
  mask_class <- labels %in% classes
  dps_test <- dps_test[mask_class]
  labels <- labels[mask_class]
  
  C <- length(classes)
  posteriors <- matrix(0, nrow = length(dps_test), ncol = C)
  
  priors <- list()
  for (c in seq(1, C))
    priors[c] <- sum(labels == classes[c]) / length(labels)
  
  for (i in 1:length(dps_test)) {
    point <- dps_test[i]
    likelihoods <- sapply(1:C, function(c) {
      predict(densities[[c]], x = point)
    })
    
    evidence <- sum(likelihoods * unlist(priors))
    
    for (c in seq(1, C)) {
      posteriors[i, c] <- (likelihoods[c] * priors[[c]]) / evidence
    }
  }
  
  labels <- match(labels, classes)
  if (C <= 2){
    roc_res <- roc(labels, posteriors[, 1])
    auc_res <- auc(roc_res)
  }
  else{
    aucs <- sapply(1:C, function(i){
      roc_res <- roc(labels == i, posteriors[, i])
      auc(roc_res)
    })
    
    auc_res <- mean(aucs)
  }
  
  return (list(auc_res = auc_res, priors_bayes = priors))
}

File: Leaspy/aux_leaspy/getAUC.py
import numpy as np
from sklearn.metrics import roc_auc_score

def getAUC(dps_test, posterior_bayes, likelihood, prior, labels, classes):
# ============================================================
# Project:    Disease progression modeling from early AD stage
# Repository: https://github.com/cplatero/preAD_DPM
# Author:     Jorge Bengoa
# Email:      j.bpinedo@alumnos.upm.es
# Institution:Universidad Politécnica de Madrid
# ------------------------------------------------------------
# Filename:    getAUC.py
# Description: Script for evaluating the Leaspy model (AUC)
#              on longitudinal neuropsychological data.
#
# Version:    1.0
# Date:       2025-05-09
# Requires:   Leaspy
# ============================================================
    dps_test = dps_test.ravel()
    mask = ~np.isnan(dps_test)
    dps_test = dps_test[mask]
    
    #One vs one AUC
    labels = labels.ravel()
    mask = np.vectorize(lambda x : x != x)(labels) # Nan != Nan
    labels = labels[~mask]
    mask_class = np.isin(labels, classes)
    
    labels = labels[mask_class]
    dps_test = dps_test[mask_class]
    
    predict_scores = posterior_bayes(likelihood, prior, dps_test)
    predict_scores = np.column_stack(list(predict_scores.values()))
    
    #Convert labels to numeric
    mapping = {nlabel: i for i, nlabel in enumerate(classes)}
    labels = np.array([mapping[nlabel] for nlabel in labels])
    
    #Remove outliers 
    mask_out = np.isnan(predict_scores[:,0]) | np.isnan(labels)
    labels = labels[~mask_out]
    predict_scores = predict_scores[~mask_out]
    
    if len(classes) == 2:
        #Binary case: the score corresponds to the probability of the class 
        #with the greater label (i.e. 2 if labels are 1 and 2)
        labels = -labels
        auc = roc_auc_score(
            labels,
            predict_scores[:, 0],
            average="macro"
        )
    else:
        #Multiclass case: the order of the class scores corresponds to the 
        #order of the labels, numerically (i.e. 1,2,3 of labels are 1,2,3)
        auc = roc_auc_score(
            labels,
            predict_scores,
            multi_class="ovo", 
            average="macro",
        )
    return auc
